interop-cats/shared/src/main/scala/zio/interop/cats.scala

```diff
@@ -35,348 +39,462 @@ abstract class CatsEffectPlatform
+
+  private def toPoll(restore: ZIO.InterruptStatusRestore) = new Poll[ZIO[R, E, *]] {
+    override def apply[T](fa: ZIO[R, E, T]): ZIO[R, E, T] = restore(fa)
+  }
 
-  override final def runCancelable[A](fa: RIO[R, A])(
-    cb: Either[Throwable, A] => effect.IO[Unit]
-  ): effect.SyncIO[effect.CancelToken[RIO[R, *]]] =
-    effect.SyncIO {
-      rts.unsafeRun {
-        ZIO.descriptor
-          .bracketExit(
-            (descriptor, exit: Exit[Throwable, A]) =>
-              ZIO.effectTotal {
-                exit match {
-                  case Exit.Failure(cause) if !cause.interruptors.forall(_ == descriptor.id) => ()
-                  case _ =>
-                    effect.IO.suspend(cb(exit.toEither)).unsafeRunAsync(_ => ())
-                }
-              },
-            _ => fa
-          )
-          .interruptible
-          .forkDaemon
-          .map(_.interrupt.unit)
-      }
+  private def toFiber[A](fiber: Fiber[E, A]) = new effect.Fiber[F, E, A] {
+    override final val cancel: F[Unit]           = fiber.interrupt.unit
+    override final val join: F[Outcome[F, E, A]] = fiber.await.map(toOutcome)
+  }
+
+  private def fiberFailure(error: E) =
+    FiberFailure(Cause.fail(error))
+
+  override def ref[A](a: A): F[effect.Ref[F, A]] =
+    ZRef.make(a).map(new ZioRef(_))
+
+  override def deferred[A]: F[Deferred[F, A]] =
+    Promise.make[E, A].map(new ZioDeferred(_))
+
+  override final def start[A](fa: F[A]): F[effect.Fiber[F, E, A]] =
+    fa.interruptible.forkDaemon.map(toFiber)
+
+  override def never[A]: F[A] =
+    ZIO.never
+
+  override final def cede: F[Unit] =
+    ZIO.yieldNow
+
+  override final def forceR[A, B](fa: F[A])(fb: F[B]): F[B] =
+    fa.either *> fb
+
+  override final def uncancelable[A](body: Poll[F] => F[A]): F[A] =
+    ZIO.uninterruptibleMask(body.compose(toPoll))
+
+  override final val canceled: F[Unit] =
+    ZIO.interrupt
+
+  override final def onCancel[A](fa: F[A], fin: F[Unit]): F[A] =
+    fa.onExit {
+      case Exit.Failure(cause) =>
+        cause.failureOrCause.fold(
+          _ => ZIO.unit,
+          _ => fin.ignore
+        )
+      case _ => ZIO.unit
     }
 
-  override final def toIO[A](fa: RIO[R, A]): effect.IO[A] =
-    effect.ConcurrentEffect.toIOFromRunCancelable(fa)(this)
+  override final def memoize[A](fa: F[A]): F[F[A]] =
+    fa.memoize
+
+  override final def racePair[A, B](fa: F[A], fb: F[B]) =
+    (fa.interruptible raceWith fb.interruptible)(
+      (exit, fiber) => ZIO.succeedNow(Left((toOutcome(exit), toFiber(fiber)))),
+      (exit, fiber) => ZIO.succeedNow(Right((toFiber(fiber), toOutcome(exit))))
+    )
+
+  override final def both[A, B](fa: F[A], fb: F[B]): F[(A, B)] =
+    fa zipPar fb
+
+  override final def guarantee[A](fa: F[A], fin: F[Unit]): F[A] =
+    fa.ensuring(fin.orDieWith(fiberFailure))
+
+  override final def bracket[A, B](acquire: F[A])(use: A => F[B])(release: A => F[Unit]): F[B] =
```

Comment:
I think we probably need to suppress the failure here and in `guarantee` to be consistent with `onCancel`.

Reply:
Makes sense 👍